### DISCUSSION TITLE
Interop with third party code that throws error

### DISCUSSION THREAD

**Discussion by shaunek-hero** [true-myth/true-myth#798] [Q&A] [answered]
I've taken a look at a few similar libraries and a feature I see in some is a way to interoperate between internal application code that fully uses `Result`/`Maybe` and third party code that may throw an error. Is there a way to do this with true-myth?

Just to provide a couple examples of what I am talking about:
1. The `ts-error-as-value` package provides a [withResult](https://www.npmjs.com/package/ts-error-as-value/v/0.4.1#withresult) function that can automatically convert a call to function that throws into a function that returns it's own Result type.
2. The `oxide.ts` package provides a [safe](https://www.npmjs.com/package/oxide.ts#safe) function that does essentially the same thing

**Comment by chriskrycho** [comment, marked as answer]
Yep! True Myth provides the Result-exception interop with [`Result.tryOr`](https://true-myth.js.org/functions/result.tryOr.html) and [`Result.tryOrElse`](https://true-myth.js.org/functions/result.tryOrElse.html). The difference between those is when the resulting error type gets allocated (including, if you have function you’re calling to build it, any other operations involved in that):

- In `tryOr` it will happen at the time you construct the `Result`, unconditionally.
- in `tryOrElse`, it will only happen if you enter the error branch.

The upside to the former is consistency and the fact that you don’t have to have an extra function floating around and/or do not need to do anything to parse the exception into a useful error type. The downside is that it will do it unconditionally. The upside to the latter form is that if you have some expensive operation, or it needs access to the exception, you can use that; but the downside is that that can actually be slower and worse in the case you don’t care about the exception value.

As a general rule of thumb, I would reach for `tryOrElse`, because I generally want to produce a meaningful error (not just a string) so that I can do further work with it at some other point in the chain.

For interop with `Maybe` and nullable types, there is [`Maybe.wrapReturn`](https://true-myth.js.org/functions/maybe.wrapReturn.html), which lets takes a function which will produce some nullable type and produces a function which will produce a `Maybe` instead—basically, if you do `let wrapped = Maybe.wrapReturn(theFn)`, calling `wrapped` will just do `Maybe.of(theFn())` under the hood. The eager version of that, then, is to call [`Maybe.of`](https://true-myth.js.org/functions/maybe.of.html) yourself directly.

We do not have the exact equivalent of `wrapReturn` for `Result`, but you can build it out of `tryOrElse` quite directly:

```ts
let safeParse = (source: string) => Result.tryOrElse(
  (error) => ({ type: 'parse-error', original: error, source }),
  JSON.parse(source),
);
```

(See [this playground][play] for a type-checking example!)

[play]: https://www.typescriptlang.org/play/?moduleResolution=99&target=99&jsx=0&module=199#code/JYWwDg9gTgLgBAJQKYGcCuAbGAaOBvOGKATwHkoBRDFJOAXzgDMoIQ4ByItJAWhGJgALAPRRUmGOwDcAKBkYk8FAENGSAArKoNOAF44AChQQ0UAMZIAXHBRFgAOwDmASj0A+QiXJUaBmXEMkKBYoazR7AGt7CAB3e1ddDwMCGGIwKw4wLRoeIJD2XGhgRwdlDGs86FxjUwt6Z2x-OAApAGVSADkAOiztJCMTcyQGmWdZeUU4MXQsPRtVDWz+gCI8OmWxoA

We should think about adding a built-in for that, too!

I’m closing this as answered, but feel free to chat further here!
